File: DarkOrbit 9.0/DarkOrbit 9.0 Emulator/src/simulator/netty/serverCommands/AbilityEffectActivationCommand.java
package simulator.netty.serverCommands;

import java.io.DataOutputStream;
import java.io.IOException;
import java.util.Vector;

import simulator.netty.ServerCommand;

/**
 Created by LEJYONER on 13/01/2018.
 */

public class AbilityEffectActivationCommand
        implements ServerCommand {

    public static int ID = 12237;

    public Vector<Integer> targetIds;   
    public int selectedAbilityId = 0;    
    public int activatorId = 0;

    public AbilityEffectActivationCommand(int param1, int param2, Vector<Integer> param3) {
        this.selectedAbilityId = param1;
        this.activatorId = param2;
        this.targetIds = param3;
    }

    public void write(DataOutputStream param1) {
        try {
            param1.writeShort(ID);
            this.writeInternal(param1);
        } catch (IOException e) {
        }
    }

    protected void writeInternal(DataOutputStream param1) {
        try {
            param1.writeInt(this.activatorId << 7 | this.activatorId >>> 25);
            param1.writeInt(this.selectedAbilityId >>> 2 | this.selectedAbilityId << 30);
            param1.writeInt(this.targetIds.size());
            for (Integer i : this.targetIds)
            {
               param1.writeInt(i >>> 16 | i << 16);
            }
            param1.writeShort(28313);
        } catch (IOException e) {
        }
    }
}

File: DarkOrbit 10.0/Net/netty/commands/AbilityEffectActivationCommand.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Ow.Utils;

namespace Ow.Net.netty.commands
{
    class AbilityEffectActivationCommand
    {
        public const short ID = 2643;

        public static byte[] write(int selectedAbilityId, int activatorId, List<int> targetIds)
        {
            var param1 = new ByteArray(ID);
            param1.writeInt(activatorId >> 4 | activatorId << 28);
            param1.writeInt(targetIds.Count);
            foreach(var _loc2_ in targetIds)
            {
                param1.writeInt(_loc2_ >> 6 | _loc2_ << 26);
            }
            param1.writeInt(selectedAbilityId << 7 | selectedAbilityId >> 25);
            param1.writeShort(-3781);
            return param1.ToByteArray();
        }
    }
}
